addons/bus/models/bus.py

```diff
@@ -94,40 +94,43 @@ def _gc_messages(self):
         domain = [('create_date', '<', timeout_ago)]
         return self.sudo().search(domain).unlink()
 
-    @api.model
-    def _sendmany(self, notifications):
-        channels = set()
-        values = []
-        for target, notification_type, message in notifications:
-            channel = channel_with_db(self.env.cr.dbname, target)
-            channels.add(channel)
-            values.append({
-                'channel': json_dump(channel),
-                'message': json_dump({
-                    'type': notification_type,
-                    'payload': message,
-                })
-            })
-        self.sudo().create(values)
-        if channels:
+    def _ensure_hooks(self):
+        if 'bus.bus.values' not in self.env.cr.precommit.data:
+            @self.env.cr.precommit.add
+            def create_bus():
+                self.sudo().create(self.env.cr.precommit.data.pop('bus.bus.values'))
+
+        if 'bus.bus.channels' not in self.env.cr.postcommit.data:
             # We have to wait until the notifications are commited in database.
             # When calling `NOTIFY imbus`, notifications will be fetched in the
             # bus table. If the transaction is not commited yet, there will be
             # nothing to fetch, and the websocket will return no notification.
             @self.env.cr.postcommit.add
             def notify():
+                payloads = get_notify_payloads(list(self.env.cr.postcommit.data.pop('bus.bus.channels')))
                 with odoo.sql_db.db_connect('postgres').cursor() as cr:
-                    query = sql.SQL("SELECT {}('imbus', %s)").format(sql.Identifier(ODOO_NOTIFY_FUNCTION))
-                    payloads = get_notify_payloads(list(channels))
                     if len(payloads) > 1:
                         _logger.info("The imbus notification payload was too large, "
                                      "it's been split into %d payloads.", len(payloads))
                     for payload in payloads:
-                        cr.execute(query, (payload,))
+                        cr.execute(SQL(
+                            "SELECT %s('imbus', %s)",
+                            SQL.identifier(ODOO_NOTIFY_FUNCTION),
+                            payload,
+                        ))
 
     @api.model
-    def _sendone(self, channel, notification_type, message):
-        self._sendmany([[channel, notification_type, message]])
+    def _sendone(self, target, notification_type, message):
```

Comment:
Note that ``_sendone`` makes no sense anymore when there is no ``_send_multi``; moreover it does not send, just add to commit hooks. In the end what we have is more a ``_sent_to_queue``, something like that.

Reply:
Maybe one of those?
* `_notify`
* `_bus_notify`
* `_bus_send`

Reply:
You are already on bus model, no need to repeat 'bus' I think (more for a generic method from a mixin, allowing to know we are doing stuff on bus even if we are on another model). ``_add_to_queue`` ? As same, it does not really _notify, but I prefer _notify to _send (well ok, bike shedding, tout ça tout ça ^^ ) .

Reply:
No more security warnings :see_no_evil: :magic_wand: :cloud: